core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/TupleToBatchConverter.scala

```diff
@@ -24,8 +27,26 @@ class TupleToBatchConverter(
 ) {
   private val partitioners = mutable.HashMap[LinkIdentity, Partitioner]()
 
+  /**
+    * Used to return the workload samples of the next operator's workers to the controller.
+    */
+  def getWorkloadHistory(): mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]] = {
+    var samples: mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]] = null
+    partitioners.values.foreach(partitioner => {
+      if (partitioner.isInstanceOf[ParallelBatchingPartitioner]) {
+        // Reshape only needs samples from workers that shuffle data across nodes
+        samples = partitioner.asInstanceOf[ParallelBatchingPartitioner].getWorkloadHistory()
+      }
+    })
+    if (samples == null) {
+      samples = new mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]]
+    }
+    samples
```

Comment:
I do not understand this logic, please correct me if I am wrong. You are scanning for each partitioner, to see if it is an instance of `ParallelBatchingPartitioner`. If found, you retrieve the history from it. The intention is clear.

However what if there are multiple `ParallelBatchingPartitioner`s? the later ones overwrites the previous ones' sample. do you want to break once found the first one?

```suggestion
    var samples: mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]] =  new mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]]
    breakable {
    partitioners.values.foreach(partitioner => {
      if (partitioner.isInstanceOf[ParallelBatchingPartitioner]) {
        // Reshape only needs samples from workers that shuffle data across nodes
        samples = partitioner.asInstanceOf[ParallelBatchingPartitioner].getWorkloadHistory()
        break
      }
    })
    }
    samples
```

Reply:
You raised a good point. Actually we shouldn't just do the first or the last in the partitioners. If there are many partitioners, all samples should be collected. I have changed the code so that we return `ArrayBuffer[mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]]]`